### DISCUSSION TITLE
locking to prevent creation of multiple stripe customers for a single user

### DISCUSSION THREAD

**Discussion by mpvosseller** [pay-rails/pay#857] [General] [answered]
I ran into an issue where two simultaneous requests caused two stripe customers to be created for a single user.

Seems like a classic multi-threading race condition in this method: https://github.com/pay-rails/pay/blob/39218494751fc021e1df89c14bcbd8812f962945/lib/pay/stripe/billable.rb#L51C1-L57C12

Example:

- Request 1: Does this user have a stripe customer yet? No, I'll create one
- Request 2: Does this user have a stripe customer yet? No, I'll create one
- Request 1: Create stripe customer and save the id
- Request 2: Create stripe customer and save the id (overwriting the previous one)

Do you think solving this (with some locking mechanism) should be the responsibility of the pay gem library or the application?

Thank you!

**Comment by excid3** [comment, marked as answer]
@mpvosseller yeah, that could definitely happen here. It would probably be wise to add a lock for all payment processors in these methods. 👍
